core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -105,16 +105,31 @@ class ReplicaFetcherThread(name:String,
       leaderEndOffset
     } else {
       /**
-       * The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
+       * If the leader's log end offset is greater than follower's log end offset, there are two possibilities:
+       * 1. The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
        * start offset because the leader has deleted old logs (log.logEndOffset < leaderStartOffset).
+       * 2. When leader migration occurs, there is a slight chance that the old leader's high watermark is greater than
+       * the new leader's log end offset, so an OffsetOutOfRangeException was thrown. After that some more messages are
+       * produced to the leader. When the follower was trying to handle the exception and query the log end offset of
```

Comment:
Yeah... I think it perhaps is not easy to explain. Actually the follower can be either A or C. Because both of them are possible to have their high watermark greater than B's Log end offset. 

The rationale behind this handling offset out of range is to figure out whether an offset OST the follower is trying to fetch was too big or to small. The way we check this is to query the log end offset (LEO) on the leader. If OST is greater than LEO it is too big, otherwise it is too small. 
However, there is some time between the follower receives the OffsetOutOfRange exception and starts to handle it. That means when the follower got the exception first time, OST might be greater than leader's LEO. But when the follower starts to handle the exception and query the LEO again, OST is already smaller than leader's LEO - because the leader might already got some new messages appended, the offset perhaps is no longer too big. In this case, current code would say OST was too small and truncate all the logs, which is a bug.